### DISCUSSION TITLE
Swap A and B in the menu (Nintendo Switch)

### DISCUSSION THREAD

**Discussion by locksoft** [diasurgical/DevilutionX#6923] [General]
First of all great, great app! I love Diablo 1 and it's wonderful being able to play it again on my Switch. <3

I've seen that all in game buttons can be personalized in the [Padmapping] section of Diablo.ini. However I'd like to have the same thing for the main menu. On the Switch A is main option and B is cancel, so I'd like to have the same, as I always confuse and do the opposite in the menu and while talking with NPCs in town. Is it possible?

Thanks. :)

**Comment by julealgon** [comment]
It would be a best practice not only to offer the option, but also to make it the default on the Switch to match the system's standards.

@locksoft not sure if you are aware of this, but you can also remap buttons in the Switch system itself. This could be a temporary workaround for you, depending on your usage scenario.

**Reply by locksoft** [reply]
Yes, I know, thanks. But it only works "in game". It has no effect on the menu.

**Reply by julealgon** [reply]
@locksoft I think you misunderstood my comment. You should be able to remap the Switch's native buttons in the Switch system menu. I'm not talking about the button remapping that DevilutionX offers.

The one configured at the system level, in the controllers menu, applies globally.

**Reply by locksoft** [reply]
Ahhh... sorry, clear now. :)
Hmmm... Nice idea, but too complex: then I'll be ok with Diablo and mad about all other Switch software. :D
But thanks. :)

**Comment by AJenbo** [comment]
In the diablo.ini config file you can overwrite the controller mapping for the game:
```ini
[Controller]
Mapping=
```
This is a bit of a workaround, for one thing I'm not sure what the default value should be for the switch controller. We do intent to try and correct it by the time 1.6.0 rolls out, but that could be a while.

**Reply by locksoft** [reply]
Thanks, but how exactly to use it?
The Wiki says about a SDL_CONTROLLERCONFIG mapping file, and it should be created connecting a controller to the PC. But I'm using Switch controllers, I'm not sure about what to do.
Thanks.

**Comment by Nephiel** [comment]
I made a [fork](https://github.com/Nephiel/devilutionX) to swap A/B behavior in menus (also X/Y in quest log, spellbook and character deletion) just for a Switch build. It's not a good solution because it is still hard-coded, and breaks support with other platforms/controller layouts. But it might be turned into a user-customizable "Swap A/B in menus" option.

**Reply by locksoft** [reply]
Cool, thanks. I’m going to test it. :)

**Reply by locksoft** [reply]
Hi Nephiel, I tested your fork for a few hours and it works perfectly, thanks. :)
It’s a bit confusing at first because, when I tried to swap A and B buttons in game (I prefer shooting with A), their label was swapped as well (being hard coded). But it’s something I had to do only once.

Your compiled binary also solved a big issue in the original release, this:

https://github.com/diasurgical/devilutionX/issues/6985

Now, turning off and on the Switch, doesn’t interfere with sounds anymore and everything works smoothly.
Maybe you can talk with @AJenbo about it, so he may “fix” regular release too?

**Reply by locksoft** [reply]
I made a quick (and bad) video about the sound problem, so maybe it's more clear:

https://github.com/user-attachments/assets/d8093c36-1379-4d43-872e-53cfae0ee52f

Sorry, I had to super-compress it because the attachment limit is 10 MB.

**Reply by DakkJaniels** [reply]
Isn't the fork based on master branch? Does this issue exist there?

**Reply by StephenCWills** [reply]
Yes, the issue exists in master. https://github.com/diasurgical/devilutionX/issues/5467#issuecomment-1926612938

I had mistakenly believed the Switch port of SDL2 would report all connected gamepads as Nintendo gamepads, but I was wrong. We need to force Nintendo layout on the Switch.

**Reply by DakkJaniels** [reply]
> Yes, the issue exists in master. https://github.com/diasurgical/devilutionX/issues/5467#issuecomment-1926612938
> 
> 
> 
> I had mistakenly believed the Switch port of SDL2 would report all connected gamepads as Nintendo gamepads, but I was wrong. We need to force Nintendo layout on the Switch.

Sorry, I was referring to the sound issue that was brought up in the post previous to mine. I should have been more clear.

**Reply by locksoft** [reply]
> Sorry, I was referring to the sound issue that was brought up in the post previous to mine. I should have been more clear.

The one I’ve shown is the last stable version (1.5.2).

**Reply by Nephiel** [reply]
I'm glad that sound issue seems fixed, but it was completely unintentional, just a happy accident. Nothing sound-related was changed in the source. But I did use `dkp-pacman` to update dev libraries before building, as per the Switch build instructions. Maybe that was fixed in those at some point after the 1.5.2 release?

Do you have the same audio options (resample rate, buffer size, resampler...) configured in both cases?

**Reply by locksoft** [reply]
> Do you have the same audio options (resample rate, buffer size, resampler...) configured in both cases?

Yes, I made a copy of the game folder and saved your binary inside. So it should have the same settings.

**Reply by Nephiel** [reply]
The fork was based on master `1.6.0-dev` at https://github.com/diasurgical/devilutionX/commit/112e113201a47dbca1c39d7684e25f01cfa89ac7 so that must have been fixed at some point between release `1.5.2` at https://github.com/diasurgical/devilutionX/commit/7ea712428ebf3df5da58e107cc40e8db3825bebc and https://github.com/diasurgical/devilutionX/commit/112e113201a47dbca1c39d7684e25f01cfa89ac7. Either in master source, or in the libraries used for the Switch build.

**Reply by locksoft** [reply]
It could be. I tested a 1.6.0 beta some time ago and it had the same audio issue, but I can’t find a link to try the last version now (I’ll never learn how to navigate the GitHub website properly). :)
